Documentation for Marker Clustering (#282)
sirensolutions/enhanced_tilemap: docs/modules/siren-investigate/pages/enhanced_tilemap.adoc

image::add_wfs_layer_button.png[image]
This allows for point, linestring (including multilinestring) and polygon (including multipolygon) types to be rendered onto the the Enhanced Coordinate Map. Polygons can be clicked on for geo filter creation

[NOTE]
==== 

> Point layers from WFS sources will render single document features rather than the <<Marker Clustering>> support for other point layer types

Reviewer: Are you saying that only single document features are available when point layers are using WFS server sources, while marker clustering allows multiple document features?

> Point layers from WFS sources will render single document features rather than the <<Marker Clustering>> support for other point layer types

Author: ```suggestion
Point layers from WFS sources will not <<Marker Clustering, cluster>>. Single document features are are drawn```

====

See image:15d88cecd8dd9c.png[image] section above for description on *Layer Name*, *Url* and *WFS Layers* fields. For details on setting up a WFS server, see the <<Getting started with GeoServer>> guide. Additional fields specific for editing WFS layers:

  * *Styling* - Set color in Hex value form and specify the size of the
  marker to display on map
  * *Output Format* - The format that your spatial server is capable of responding with. The format for ArcGis Server is GeoJSON, and the format for Geoserver is JSON.
  * *Popup Content* - To configure a pop-up tooltip, create a comma-separated list of fields in the properties object of your GeoJSON features. For example, the parameter `City_name,Pop_est` adds the city name and the estimated population fields to each feature that is added to the map. *Note*: This parameter is case-sensitive. 

==== Marker Clustering

> Point layers from <<Point of Interest layers, Point of Interest>> and <<Loading Stored Layers into Elasticsearch, Stored Layer>> sources use Marker Clustering. This approach means that all documents on the current map canvas are represented at once. 

Reviewer: ```suggestion
Marker clustering is a method that allows all documents on the current map canvas to be represented at once. It is used by <<Point of Interest layers, Point of Interest>> layers and <<Loading Stored Layers into Elasticsearch, Stored Layer>> sources. 
```

Reviewer: @corrigancd I don't think the first link to Point of Interest layers is working. Don't these links work for anchored headings only? I couldn't find the heading for "Point of Interest layers", so please remove the link syntax if it's not going to work.

===== How does it work?

> Two queries are used each time this layer type is re-rendered. Initially, a geohash aggregation query is fired. Based on this, geohashes that have the least number of documents present are selected up to the point that the configured limit would be exceeded. These selected geohashes are removed and Geo-Filters are created for these instead. A document query is then fired to retrive individual documents within these areas. The configured aggregation <<The Options Tab, Map Collar Scale>> is used for both queries.

Reviewer: ```suggestion
Each time this layer type is re-rendered, the following two queries are used: 
* Firstly, a geo-hash aggregation query is initiated. Based on this query, geo-hashes that have the least number of documents present are selected up to the limit that is configured. These selected geo-hashes are removed and geo-filters are created for these instead. 
* Secondly, a document query is initiated to retrieve individual documents in these areas. 
The configured <<The Options Tab, Map Collar Scale>> aggregation is used for both queries.
```

===== How does it look?

> The result is demonstrated below, the cluster can be distinguished from the number to the right (391). When a cluster is hovered over, the corresponding geohash appears. When a marker corresponding to one document is hovered over, a popup, if configured (see <<Configuring Stored Layers, Stored Layers>> and <<Point of Interest layers, POI>> for respective configurations) appears.

Reviewer: ```suggestion
The result is displayed in the following image. The cluster can be distinguished by the number (391). When you hover your cursor over a cluster, the corresponding geo-hash is displayed. 
When a marker that corresponds to one document is hovered over, a pop-up window can be displayed if it is configured. For more information, see <<Configuring Stored Layers, Stored Layers>> and <<Point of Interest layers, POI>> for respective configuration steps.
```

> image:marker_clustering.png[Marker Clustering Example]

Reviewer: ```suggestion
image:marker_clustering.png[Example of marker clustering]
```

> ==== Configuring Stored Layers

Reviewer: ```suggestion
==== Configuring stored layers
```

* Point Layers
* Shape Layers (polygons and lines)

> * Heatmap Aggregation layer (if this type of Aggregation is selected, this will appear above tile layers and below all other Shape layers)

Reviewer: ```suggestion
* Heatmap aggregation layer: If this type of aggregation is selected, this will appear above tile layers and below all other shape layers.
```
